src/rpc/blockchain.cpp

```diff
@@ -650,23 +647,23 @@ UniValue getblock(const UniValue &params, bool fHelp)
             HelpExampleCli("getblock", "\"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09\"") +
             HelpExampleRpc("getblock", "\"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09\""));
 
-    LOCK(cs_main);
 
     std::string strHash = params[0].get_str();
     uint256 hash(uint256S(strHash));
-    CBlockIndex *pblockindex = NULL;
     bool fVerbose = true;
     bool fListTxns = true;
     if (params.size() > 1)
         fVerbose = params[1].get_bool();
     if (params.size() == 3)
         fListTxns = params[2].get_bool();
 
-    if (mapBlockIndex.count(hash) == 0)
+    CBlockIndex *pblockindex = LookupBlockIndex(hash);
+    if (!pblockindex)
     {
         arith_uint256 h = UintToArith256(hash);
         if (h.bits() < 65)
         {
+            LOCK(cs_main);
             uint64_t height = boost::lexical_cast<unsigned int>(strHash);
```

Comment:
I don't think chainActive.Height() needs cs_main anymore...it relies on tip() which is atomic.

Reply:
its needed for "pblockindex = chainActive[height];" just below.  

giving chainActive its own lock is the next thing to do for multithread perf I think...